// @webex/test-helper-file/dist/file.js
"use strict";

var _Object$keys = require("@babel/runtime-corejs2/core-js/object/keys");
var _Object$getOwnPropertySymbols = require("@babel/runtime-corejs2/core-js/object/get-own-property-symbols");
var _Object$getOwnPropertyDescriptor = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptor");
var _Object$getOwnPropertyDescriptors = require("@babel/runtime-corejs2/core-js/object/get-own-property-descriptors");
var _Object$defineProperties = require("@babel/runtime-corejs2/core-js/object/define-properties");
var _Object$defineProperty = require("@babel/runtime-corejs2/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.isMatchingFile = exports.isBufferLike = exports.isBlobLike = exports.fetchWithoutMagic = exports.fetch = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/slicedToArray"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/defineProperty"));
var _promise = _interopRequireDefault(require("@babel/runtime-corejs2/core-js/promise"));
var _fs = require("fs");
var _path = require("path");
var _fileType = require("file-type");
var _common = require("@webex/common");
function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; } /*!
 * Copyright (c) 2015-2020 Cisco Systems, Inc. See LICENSE file.
 */
var fetchWithoutMagic = exports.fetchWithoutMagic = function fetchWithoutMagic(filename) {
  return new _promise.default(function (resolve, reject) {
    var filepath = (0, _path.join)(__dirname, '../../../@webex/test-helper-server/static', filename);
    (0, _fs.readFile)(filepath, function (err, data) {
      if (err) {
        reject(err);
        return;
      }
      resolve(_objectSpread(_objectSpread({}, data), {}, {
        name: filename
      }));
    });
  });
};
var fetch = exports.fetch = function fetch(filename) {
  return fetchWithoutMagic(filename).then(function (data) {
    return _promise.default.all([(0, _fileType.fromFile)(data), data]);
  })
  // .then(([{mime: type}, data]) => {
  //   console.error(data, type);

  //   return {...data, type};
  // });
  .then(function (_ref) {
    var _ref2 = (0, _slicedToArray2.default)(_ref, 2),
      type = _ref2[0].mime,
      data = _ref2[1];
    return type ? _objectSpread(_objectSpread({}, data), {}, {
      type: type
    }) : _promise.default.reject(new Error('Invalid Media Type'));
  });
};
var isBufferLike = exports.isBufferLike = function isBufferLike(file) {
  return (0, _common.isBuffer)(file);
};
var isBlobLike = exports.isBlobLike = function isBlobLike(file) {
  return (0, _common.isBuffer)(file);
};
var isMatchingFile = exports.isMatchingFile = function isMatchingFile(left, right) {
  if (!isBufferLike(left)) {
    throw new Error('`left` must be a `Buffer`');
  }
  if (!isBufferLike(right)) {
    throw new Error('`right` must be a `Buffer`');
  }

  // Node 10 doesn't have Buffer#equals()
  if (left.equals) {
    return _promise.default.resolve(left.equals(right));
  }
  if (left.length !== right.length) {
    return _promise.default.resolve(false);
  }
  for (var i = 0; i < left.length; i += 1) {
    if (left[i] !== right[i]) {
      return _promise.default.resolve(false);
    }
  }
  return _promise.default.resolve(true);
};
